ampd/src/event_sub/stream.rs

```diff
@@ -0,0 +1,479 @@
+
+use crate::asyncutil::future::{with_retry, RetryPolicy};
+use crate::tm_client::TmClient;
+
+type Error = super::Error;
+type Result<T> = error_stack::Result<T, Error>;
+
+const BLOCK_PROCESSING_BUFFER: usize = 10;
+
+pub async fn blocks<T>(
+    tm_client: &T,
+    poll_interval: Duration,
+    token: CancellationToken,
+) -> Result<impl Stream<Item = Result<block::Height>> + '_>
+where
+    T: TmClient,
+{
+    latest_block_height(tm_client)
+        .await
+        .map(|latest| BlockState {
+            current: latest,
+            latest,
+        })
+        .map(|block_state| block_state.stream(tm_client, interval(poll_interval), token))
+        .map(Box::pin)
+}
+
+pub fn events<'a, T, S>(
+    tm_client: &'a T,
+    block_stream: S,
+    retry_policy: RetryPolicy,
+) -> impl Stream<Item = Result<Event>> + 'a
+where
+    T: TmClient,
+    S: Stream<Item = Result<block::Height>> + 'a,
+{
+    block_stream
+        .map(move |block_height| process_block(tm_client, block_height, retry_policy))
+        .buffered(BLOCK_PROCESSING_BUFFER)
+        .flat_map(|result| {
+            result.map_or_else(
+                |err| stream::iter(vec![Err(err)]),
+                |events| stream::iter(events.into_iter().map(Ok).collect::<Vec<_>>()),
+            )
+        })
+}
+
+async fn process_block<T>(
+    tm_client: &T,
+    block_height: Result<block::Height>,
+    retry_policy: RetryPolicy,
+) -> Result<Vec<Event>>
+where
+    T: TmClient,
+{
+    match block_height {
+        Ok(block_height) => {
+            with_retry(|| block_events(tm_client, block_height), retry_policy).await
+        }
+        Err(err) => Err(err),
+    }
+}
+
+async fn block_events<T>(tm_client: &T, block_height: block::Height) -> Result<Vec<Event>>
+where
+    T: TmClient,
+{
+    let block_results =
+        tm_client
+            .block_results(block_height)
+            .await
+            .change_context(Error::BlockResultsQuery {
+                block: block_height,
+            })?;
+
+    let begin_block_events = block_results.begin_block_events.into_iter().flatten();
+    let tx_events = block_results
+        .txs_results
+        .into_iter()
+        .flatten()
+        .flat_map(|tx| tx.events);
+    let end_block_events = block_results.end_block_events.into_iter().flatten();
+
+    let events = begin_block_events
+        .chain(tx_events)
+        .chain(end_block_events)
+        .map(|event| {
+            Event::try_from(event).change_context(Error::EventDecoding {
+                block: block_height,
+            })
+        })
+        .collect::<Result<Vec<_>>>()?;
+
+    Ok(iter::once(Event::BlockBegin(block_height))
+        .chain(events)
+        .chain(iter::once(Event::BlockEnd(block_height)))
+        .collect())
+}
+
+#[derive(Copy, Clone)]
+struct BlockState {
+    current: block::Height,
+    latest: block::Height,
+}
+
+impl BlockState {
+    async fn update<T>(mut self, tm_client: &T, interval: &mut Interval) -> Result<Self>
```

Comment:
pass in cancellation? This could hang ampd on a chain halt (e.g. upgrade) otherwise